tools/navcon-converter/navcon-converter

```diff
@@ -0,0 +1,139 @@
+# SOFTWARE.
+#
+# ===========================================================================
+
+import struct
+import sys
+
+"""
+Version 2 binary format
+
+Header version number then connection count then arrays of various types.
+
+Example:
+header count {x1 z1 y1 x2 z2 y2}[] rad[] flags[] areas[] dirs[] userids[]
+i      i      f  f  f  f  f  f     f     H       B       B      I
+
+int header;
+int count;
+union coords_t { float x1, z1, y1, x2, z2, y2 };
+coords_t[count] coords;
+float rad[count];
+unsigned short flags[count];
+unsigned byte areas[count];
+unsigned byte dirs[count];
+unsigned int userids[count];
+"""
+
+"""
+Version 3 text format
+
+Header line with prefix and integer version number
+then one connection per line, all numbers as integers.
+
+Example:
+#navcon 3
+x1 z1 y1 x2 z2 y2 rad flags areas dirs userids
+[…]
+"""
+
+current_version = 3
+
+for file_name in sys.argv[1:]:
+    header_text_prefix = "#navcon"
+    header_text_magic = struct.unpack('<i', bytes(header_text_prefix[:4], "utf-8"))[0]
+
+    file_blob = open(file_name, "rb")
+
+    header_magic = struct.unpack('<i', file_blob.read(4))[0]
+
+    if header_magic == 2:
+        format_size_dict = {
+            "f": 4,
+            "I": 4,
+            "H": 2,
+            "B": 1,
+        }
+
+        format_type_dict = {
+            "x1": "f",
+            "z1": "f",
+            "y1": "f",
+            "x2": "f",
+            "z2": "f",
+            "y2": "f",
+            "rad": "f",
+            "flags": "H",
+            "areas": "B",
+            "dirs": "B",
+            "userids": "I",
+        }
+
+        coords_list = ["x1", "z1", "y1", "x2", "z2", "y2"]
+
+        print("Converting {}".format(file_name), file=sys.stderr)
+        print("Converting from version {} to {}".format(header_magic, current_version), file=sys.stderr)
+
+        con_count = struct.unpack('<i', file_blob.read(4))[0]
+        print("Converting {} connections".format(str(con_count)), file=sys.stderr)
+
+        con_list = [{} for i in range(0, con_count)]
+
+        for key_name in format_type_dict.keys():
+            for i in range(0, con_count):
+
+                if key_name == coords_list[0]:
+                    subkey_list = coords_list
+                elif key_name in coords_list[1:]:
+                    continue
+                else:
+                    subkey_list = [key_name]
+
+                for subkey_name in subkey_list:
+                    value_format = format_type_dict[subkey_name]
+                    value_size = format_size_dict[value_format]
+
+                    value_bytes = file_blob.read(value_size)
+                    value_int = int(struct.unpack(value_format, value_bytes)[0])
+
+                    con_list[i][subkey_name] = value_int
+
+        file_blob.close()
+
+        print("{} {}".format(header_text_prefix, str(current_version)))
+
+        for con in con_list:
+            for key in con.keys():
```

Comment:
This loop could be replaced by `print(*con.values())`

Reply:
Wow, what is this dark magic???.
done.